File: research/endgame.cpp
// build_memo.cpp (revised)
// Compile: g++ -O3 -std=c++17 -o build_memo build_memo.cpp
// -----------------------------------------------------------------------------
// Generates two files:
//   1. memo.jsonl  –  every hand (≤15 cards) whose best_move is **non-trivial**
//                     ("h" = hand array, "s" = score, "m" = [len,endRank]).
//   2. sample.txt  –  up to 1 000 non-trivial hands with only hand & score for
//                     quick sanity-checking.
// After the run it prints the distribution of scores for both trivial and
// non-trivial positions as well as the counts of each category.
// -----------------------------------------------------------------------------
#include <array>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <optional>
#include <unordered_map>
#include <vector>

// ─────────────────────────────────── Constants ───────────────────────────────
constexpr int RANKS      = 13;                      // 0 = 3 … 12 = 2
constexpr int MAX_CARDS  = 15;
constexpr std::array<uint8_t, RANKS> MAX_COUNTS{4,4,4,4,4,4,4,4,4,4,4,3,1};

// ────────────────────────── Hand representation & hashing ────────────────────
using Hand = std::array<uint8_t, RANKS>;            // small, trivially copyable
struct HandHash {
    std::size_t operator()(const Hand& h) const noexcept {
        std::size_t hash = 14695981039346656037ull; // FNV-1a
        for (uint8_t c : h) {
            hash ^= c;
            hash *= 1099511628211ull;
        }
        return hash;
    }
};
struct Result {
    uint8_t                                       score;   // best score (0-15)
    std::optional<std::pair<uint8_t,uint8_t>>     move;    // (length, endRank)
};
static std::unordered_map<Hand, Result, HandHash> MEMO;

// ─────────────────────────────────── Forward decl. ───────────────────────────
Result fun(const Hand& hand);

// ───────────────────────── Hand generator (depth-first) ──────────────────────
void gen_rec(int idx, int total, Hand& hand, std::vector<Hand>& out) {
    if (idx == RANKS) { out.push_back(hand); return; }
    const uint8_t cap = MAX_COUNTS[idx];
    for (uint8_t c = 0; c <= cap; ++c) {
        int new_total = total + c;
        if (new_total > MAX_CARDS) break;           // prune
        hand[idx] = c;
        gen_rec(idx+1, new_total, hand, out);
    }
    hand[idx] = 0;                                  // restore
}

// ───────────────────────────── Utility helpers ───────────────────────────────
inline bool all_positive(const Hand& h) {
    for (uint8_t x : h) if (!x) return false;
    return true;
}

// ─────────────────────────── DP core (heuristic) ─────────────────────────────
Result fun(const Hand& hand) {
    if (auto it = MEMO.find(hand); it != MEMO.end()) return it->second;

    uint8_t out = 15;                                      // worst/best sentinel
    std::optional<std::pair<uint8_t,uint8_t>> best_move;   // none yet

    // numExtras = 4-of-kind or Ace-triple bonus cards +1 baseline
    uint8_t numExtras = 1;
    for (int i = 0; i < RANKS; ++i)
        if (hand[i] == 4 || (i == 11 && hand[i] == 3)) ++numExtras;

    // simple discard heuristic
    for (int rank = 0; rank < RANKS; ++rank) {
        if (hand[rank] == 1) {
            if (numExtras) --numExtras;
            else { out = rank + 3; break; }
        }
    }
    if (out == 15) {
        Result res{15, std::nullopt};
        MEMO.emplace(hand, res);
        return res;
    }

    // working copy
    Hand tmp = hand;

    // straight lengths 13 .. 5
    for (int len = 13; len >= 5; --len) {
        int max_start = 13 - len;
        for (int s = 0; s <= max_start; ++s) {
            bool legal = true;
            for (int j = 0; j < len; ++j) if (!hand[s+j]) { legal = false; break; }
            if (!legal) continue;
            for (int j = 0; j < len; ++j) tmp[s+j]--;
            auto sub = fun(tmp);
            for (int j = 0; j < len; ++j) tmp[s+j]++;
            if (sub.score > out) {
                out = sub.score;
                best_move = {static_cast<uint8_t>(len), static_cast<uint8_t>(s+len-1)};
                if (out == 15) goto finish;               // perfect
            }
        }
        // bottom-2 special straight
        if (hand[12] && len < 13) {
            bool legal = true;
            for (int j = 0; j < len-1; ++j) if (!hand[j]) { legal = false; break; }
            if (legal) {
                tmp[12]--; for (int j = 0; j < len-1; ++j) tmp[j]--;
                auto sub = fun(tmp);
                tmp[12]++; for (int j = 0; j < len-1; ++j) tmp[j]++;
                if (sub.score > out) {
                    out       = sub.score;
                    best_move = {static_cast<uint8_t>(len), static_cast<uint8_t>(len-2)};
                    if (out == 15) goto finish;
                }
            }
        }
    }

finish:
    Result res{out, best_move};
    MEMO.emplace(hand, res);
    return res;
}

// ───────────────────────────────────── main ──────────────────────────────────
int main() {
    // Generate universe of hands
    std::vector<Hand> hands; Hand h{}; gen_rec(0,0,h,hands);

    std::ofstream jsonl("memo.jsonl");
    std::ofstream sample("sample.txt");

    std::array<std::uint64_t,16> trivDist{};      // score freq for trivial
    std::array<std::uint64_t,16> nonTrivDist{};   // score freq for non-trivial

    std::size_t nonTrivialCnt = 0;
    const std::size_t SAMPLE_CAP = 1000;

    std::size_t processed = 0;
    for (const auto& hd : hands) {
        auto res = fun(hd);
        ++processed;
        if (res.move) {                           // non-trivial
            ++nonTrivialCnt;
            ++nonTrivDist[res.score];
            // write full JSON-line
            jsonl << "{\"h\":[";
            for (int i = 0; i < RANKS; ++i) {
                jsonl << int(hd[i]);
                if (i+1 != RANKS) jsonl << ',';
            }
            jsonl << "],\"s\":" << int(res.score)
                  << ",\"m\":[" << int(res.move->first)
                  << ',' << int(res.move->second) << "]}\n";
            // sampled plain hand+score
            if (nonTrivialCnt <= SAMPLE_CAP) {
                for (int i = 0; i < RANKS; ++i) {
                    sample << int(hd[i]) << (i+1==RANKS ? ' ' : ',');
                }
                sample << int(res.score) << '\n';
            }
        } else {
            ++trivDist[res.score];
        }
        if (processed % 100000 == 0)
            std::cerr << "processed " << processed << " hands...\n";
    }

    jsonl.close(); sample.close();

    // ───────────── print distributions ─────────────
    std::cout << "\nScore distribution (trivial vs non-trivial)\n";
    std::cout << "score\ttrivial\tnon_trivial\n";
    for (int s = 0; s <= 15; ++s)
        std::cout << s << '\t' << trivDist[s] << '\t' << nonTrivDist[s] << '\n';

    std::cout << "\nTotal hands        : " << hands.size() << '\n';
    std::cout << "Non-trivial hands  : " << nonTrivialCnt << '\n';
    std::cout << "memo.jsonl lines   : " << nonTrivialCnt << '\n';
    std::cout << "sample.txt lines   : " << std::min<std::size_t>(nonTrivialCnt,SAMPLE_CAP) << '\n';
    return 0;
}

File: research/endgame.py
#!/usr/bin/env python3
"""Build DP table for Big-2 hands ≤ 15 cards and save to memo.jsonl."""

from functools import lru_cache
import json
from pathlib import Path

# ────────────────────────────────────────────────────────────────
#   CONSTANTS
# ────────────────────────────────────────────────────────────────

# 0 = rank 3, 1 = rank 4, …, 10 = rank K, 11 = A, 12 = 2
MAX_COUNTS   = (4,)*11 + (3, 1)        # tuple is slightly faster than list
MAX_CARDS    = 15
RANKS        = range(13)
STRAIGHT_RNG = tuple(range(5, 14))     # possible straight lengths 5‥13

# ────────────────────────────────────────────────────────────────
#   HAND GENERATOR  (depth-first with pruning)
# ────────────────────────────────────────────────────────────────

def generate_hands():
    """Yield every legal 13-tuple with total ≤ 15."""
    hand = [0]*13
    def rec(idx: int, total: int):
        if idx == 13:
            yield tuple(hand)           # convert once at the leaf
            return
        cap = MAX_COUNTS[idx]
        for c in range(cap+1):
            new_total = total + c
            if new_total > MAX_CARDS:   # prune branch
                break
            hand[idx] = c
            yield from rec(idx+1, new_total)
        hand[idx] = 0                   # reset before back-tracking
    yield from rec(0, 0)

# ────────────────────────────────────────────────────────────────
#   DP  (fun) — now using lru_cache and micro-optimisations
# ────────────────────────────────────────────────────────────────

@lru_cache(maxsize=None)
def fun(hand: tuple[int, ...]) -> tuple[int, tuple | None]:
    """
    Returns (best_score, best_move) for a given hand.
    best_score == 15  ⇢  terminal / optimal
    best_move   == whatever encoding you like (kept same as your draft)
    """
    # ---------- fast checks ----------
    out        = 15    # pessimistic upper bound
    best_move  = None
    num_extras = 1 + sum(1 for i, c in enumerate(hand)
                         if c == 4 or (i == 11 and c == 3))

    # discard triples → doubles → bombs heuristic
    for rank in RANKS:
        if rank == 1:          # special-case “4s beat 3s” rule in your draft
            if num_extras:
                num_extras -= 1
            else:
                out = rank + 3
                break

    if out == 15:
        return 15, None        # already optimal

    # ---------- try straights (length 13‥5) ----------
    hand_list = list(hand)     # mutate cheaply inside loops
    for length in STRAIGHT_RNG[::-1]:          # 13,12,…,5
        max_start = 13 - length
        for start in range(max_start+1):
            if all(hand[start+j] for j in range(length)):
                for j in range(length):
                    hand_list[start+j] -= 1
                res, _ = fun(tuple(hand_list))
                for j in range(length):
                    hand_list[start+j] += 1    # restore
                if res > out:
                    out, best_move = res, (length, start+length-1)
                    if out == 15:
                        return 15, best_move

        # bottom-2 straight (special rule in your draft)
        if hand[12] and length >= 5:
            if all(hand[j] for j in range(length-1)):
                hand_list[12] -= 1
                for j in range(length-1):
                    hand_list[j] -= 1
                res, _ = fun(tuple(hand_list))
                hand_list[12] += 1
                for j in range(length-1):
                    hand_list[j] += 1
                if res > out:
                    out, best_move = res, (length, length-2)
                    if out == 15:
                        return 15, best_move

    # ---------- straight of 13 covering all ranks ----------
    if all(hand):
        for j in RANKS:
            hand_list[j] -= 1
        res, _ = fun(tuple(hand_list))
        for j in RANKS:
            hand_list[j] += 1
        if res > out:
            out, best_move = res, (13, 12)

    return out, best_move

# ────────────────────────────────────────────────────────────────
#   DRIVER — stream results to disk
# ────────────────────────────────────────────────────────────────

def main(out_path: str | Path = "memo.jsonl"):
    out_path = Path(out_path).expanduser()
    n_states = 0
    with out_path.open("w") as fh:
        for hand in generate_hands():
            score, move = fun(hand)
            fh.write(json.dumps({"h": hand, "s": score, "m": move}) + "\n")
            n_states += 1
            if n_states % 10_000 == 0:
                print(f"processed {n_states:,} hands...")

    print(f"Finished: {n_states:,} hands written to {out_path}")

if __name__ == "__main__":
    main()
